### PR TITLE
mirror: Add feature flag to allow initial user creation (PROJQUAY-1926)

### PR DESCRIPTION
**Pull request by jonathankingfc**
- Users can use FEATURE_INIT_USER to allow an initial user to be created.
- Credentials can be passed through `INIT_USERNAME` and `INIT_PASSWORD` otherwise defaults will be used

### CODE REVIEW

**Comment by LiZhang19817** [comment]
Will this feature "FEATURE_INIT_USER" be exposed on quay config editor?

**Comment by alecmerdler** [comment]
Agreed with @LiZhang19817, this PR should be accompanied by a subsequent PR to `config-tool`.

File: Makefile [hunk 1]
```diff
@@ -195,10 +195,10 @@ local-dev-build:
 .PHONY: local-dev-up
 local-dev-up:
 	make local-dev-clean
-	docker-compose up -d redis
-	docker-compose up -d quay-db
+	docker-compose up  --build -d redis
```
Comment (alecmerdler) [inline Makefile:198]: Can you explain what this does?

File: local-dev/stack/config.yaml [hunk 2]
```diff
@@ -30,7 +30,7 @@ FEATURE_REPO_MIRROR: false
 FEATURE_REQUIRE_TEAM_INVITE: true
 FEATURE_RESTRICTED_V1_PUSH: false
 FEATURE_SECURITY_NOTIFICATIONS: true
-FEATURE_SECURITY_SCANNER: true
```
Comment (alecmerdler) [inline local-dev/stack/config.yaml:33]: Is disabling the security scanner by default an intended part of this PR?

File: local-dev/stack/config.yaml [hunk 3]
```diff
@@ -69,3 +69,6 @@ USER_EVENTS_REDIS:
   host: quay-redis
   port: 6379
 USE_CDN: false
+FEATURE_INIT_USER: true
```
Comment (alecmerdler) [inline local-dev/stack/config.yaml:72]: Prefer `FEATURE_INITIAL_USER`, `INITIAL USERNAME`, and `INITIAL_PASSWORD`.

**Review by alecmerdler — CHANGES_REQUESTED** [review]
This deserves further discussion as it involves database rows defined by `config.yaml`. 

- What happens if this user is deleted? When Quay restarts, is the user recreated?
- Should there be the same password requirements on the initial user password as on the rest of Quay's users? Will that cause issues if they are different?
- What happens if the `config.yaml` is modified to remove this initial user, but it exists previously? Is it deleted?

**Comment by alecmerdler** [comment]
I think this is significant enough of a feature to deserve an [enhancement proposal](https://github.com/quay/enhancements), especially to capture the motivations and design.

File: data/model/user.py [hunk 4]
```diff
@@ -111,6 +111,34 @@ def create_user(
...
+):
+    """
+    Creates an initial user, if none are in the database.
+    """
+
+    if config.app_config.get("AUTHENTICATION_TYPE") != "Database":
+        logger.debug("Cannot create init user in a non-database auth system")
+        return
+
+    if bool(list(User.select().limit(1))):
```
Comment (kleesc) [inline data/model/user.py:127]: We've generally favored running the query and catching the "DoesNotExist" exception when testing for existence.
e.g 
```
try:
  User.select().limit(1).get()
except UserDoesNotExist:
  logger.debug("Cannot initialize user in a non-empty database")
  return
```

File: data/model/user.py [hunk 5]
```diff
@@ -111,6 +111,34 @@ def create_user(
     return created
 
 
+def create_init_user(
```
Comment (kleesc) [inline data/model/user.py:114]: Can we add a test for this? Just to test the new constraints we're adding to `create_user`.

**Review by kleesc — CHANGES_REQUESTED** [review]

**Comment by kleesc** [comment]
@jonathankingfc Is this a duplicate of #771 ?

**Comment by kleesc** [comment]
> I think this is significant enough of a feature to deserve an [enhancement proposal](https://github.com/quay/enhancements), especially to capture the motivations and design.

@alecmerdler @jonathankingfc Agree. I'm not sure what the scope of this needs to be. Are we just looking to get access to the registry API, or also to Quay's API. AFAIK the latter also requires an application/org to be able to create an oauth token to be able to use on behalf of that first user.

**Comment by jonathankingfc** [comment]
Closing in favor of https://github.com/quay/quay/pull/771

**Comment by jonathankingfc** [comment]
Closing in favor of https://github.com/quay/quay/pull/771

**Pull request closed without merging**